rmw_connext_cpp/src/functions.cpp

```diff
@@ -169,48 +167,41 @@ ros_middleware_interface::PublisherHandle create_publisher(const ros_middleware_
         DDS_BOOLEAN_FALSE);
     if (status != DDS_RETCODE_OK)
     {
-        printf("  add_property() could not add qos property\n");
-        throw std::runtime_error("could not add qos property");
+        rmw_set_error_string("  add_property() could not add qos property");
+        return NULL;
     }
-    std::cout << "  create_publisher() limit preallocated sample size" << std::endl;
 
-    std::cout << "  create_publisher() create data writer" << std::endl;
     DDSDataWriter* topic_writer = dds_publisher->create_datawriter(
         topic, datawriter_qos,
         NULL, DDS_STATUS_MASK_NONE);
 
 
-    std::cout << "  create_publisher() build opaque publisher handle" << std::endl;
     CustomPublisherInfo* custom_publisher_info = new CustomPublisherInfo();
     custom_publisher_info->topic_writer_ = topic_writer;
     custom_publisher_info->callbacks_ = callbacks;
 
-    ros_middleware_interface::PublisherHandle publisher_handle = {
-        _rti_connext_identifier,
-        custom_publisher_info
-    };
-    return publisher_handle;
+    rmw_publisher_t * publisher = new rmw_publisher_t;
+    publisher->implementation_identifier = _rti_connext_identifier;
+    publisher->data = custom_publisher_info;
+    return publisher;
 }
 
-void publish(const ros_middleware_interface::PublisherHandle& publisher_handle, const void * ros_message)
+rmw_ret_t
+rmw_publish(const rmw_publisher_t * publisher, const void * ros_message)
 {
-    //std::cout << "publish()" << std::endl;
-
-    if (publisher_handle._implementation_identifier != _rti_connext_identifier)
+    if (publisher->implementation_identifier != _rti_connext_identifier)
     {
-        printf("publisher handle not from this implementation\n");
-        printf("but from: %s\n", publisher_handle._implementation_identifier);
-        throw std::runtime_error("publisher handle not from this implementation");
+        rmw_set_error_string("publisher handle not from this implementation");
+        // rmw_set_error_string("but from: %s\n", publisher->implementation_identifier);
```

Comment:
`rmw_set_error_string` does not accept format strings, but it'd be a nice addition IMO.